text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,152 @@
+# `upgrade-bootstrapper`
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0032](https://github.com/casperlabs/ceps/pull/0032)
+
+The `upgrade-bootstrapper` is a new binary which will be responsible for launching the `casper-node` binaries.  It will
+handle transitions between network versions by identifying when a current `casper-node` execution ceases and starting
+a new version of the `casper-node` binary.
+
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The `upgrade-bootstrapper` will be a relatively simple Rust binary which will run the `casper-node` binary as a child
+process.  It will take a single optional argument: the path to the `config.toml` for the `casper-node` binary to be run.
+
+Under normal circumstances, the following steps will be performed:
+
+1. the `upgrade-bootstrapper` is started
+2. if the path to the node's `config.toml` was provided via an arg, the `upgrade-bootstrapper` writes the path value to
+   a hard-coded, known location (if writing fails, it panics).  Otherwise it reads the path value from the known
+   location.  If the path cannot be read from the known location, it panics
+3. the `upgrade-bootstrapper` spawns a child process, running the `casper-node` binary in validator mode, passing it the
+   path to its `config.toml` as an arg
+4. while they are running, before the network reaches the upgrade point, the next version of the `casper-node` gets
+   installed, but not run.  The binary is always named `casper-node-next` and its config named `config-next.toml`.  Each
```

Comment:
Alternatively, we could call it `casper-node-<version>`, then multiple upgrades could be scheduled at the same time?

Reply:
I considered this, but wanted to try and keep the process as simple as possible for now.

The only use case I could think of for that was where a user has missed one or more intermediate updates.  So, e.g. their node was last running at version 1.0.0, but they've missed 2.0.0 and the network is currently on 3.0.0.

We could have the `upgrade-bootstrapper` search the install folder for the appropriate next package, but that adds a fair bit of complexity which I'm not sure is worthwhile.  In this case, the `upgrade-boostrapper` would probably only want to run the migration step of the v2.0.0 binary since there'd be no point running the node at v2.0.0 since it shouldn't be able to communicate with any of the peers at v3.0.0.  However, this might not be the case for a minor or patch level change.

If we think this overhead is indeed worthwhile, I can spend some further time updating the doc to try to cover these various flows.  However, I'd quite like to get general agreement on the current approach before spending time on this, in case it's wasted effort :)